### PR TITLE
Fix possible SIGSEV after producer close in modelgateway

### PR DESCRIPTION
**Pull request by cliveseldon**
[Its not allowed to use a producer after or during its closed](https://github.com/confluentinc/confluent-kafka-go/issues/909).
This PR wraps a lock around produce to ensure safe multi-goroutine use.

### CODE REVIEW

**Comment by seldondev** [comment]
[APPROVALNOTIFIER] This PR is **NOT APPROVED**

This pull-request has been approved by:
To complete the [pull request process](https://git.k8s.io/community/contributors/guide/owners.md#the-code-review-process), please assign 
You can assign the PR to them by writing `/assign ` in a comment when ready.

The full list of commands accepted by this bot can be found [here](https://jenkins-x.io/v3/develop/reference/chatops/?repo=SeldonIO%2Fseldon-core).

<details open>
Needs approval from an approver in each of these files:

- **[OWNERS](https://github.com/SeldonIO/seldon-core/blob/v2/OWNERS)**

Approvers can indicate their approval by writing `/approve` in a comment
Approvers can cancel approval by writing `/approve cancel` in a comment
</details>

File: scheduler/pkg/kafka/gateway/consumer.go
Comment (sakoush) [inline scheduler/pkg/kafka/gateway/consumer.go:177]: I understand that this is technically not part of this PR but it feels odd to have a producer in a InferKafkaConsumer struct. Not sure if there is a refactor that could reduce the confusion.

File: scheduler/pkg/kafka/gateway/consumer.go
Comment (sakoush) [inline scheduler/pkg/kafka/gateway/consumer.go:293]: Related to my comment about a producer in a consumer struct, there are two running flags that confused me initially thinking they should be only one of them. At least rename to reflect what's the flag concern?

**Review by sakoush — COMMENTED** [review]
Left some comments about readability.

File: scheduler/pkg/kafka/gateway/consumer.go
Comment (agrski) [inline scheduler/pkg/kafka/gateway/consumer.go:190]: :question: Is it allowed to close a producer (or consumer) multiple times?  If not, we should be checking here if it's already closed by the time this method takes the lock.

File: scheduler/pkg/kafka/gateway/consumer.go
Comment (agrski) [inline scheduler/pkg/kafka/gateway/consumer.go:183]: :question: No `logger`?  The `kc` instance has a logger that presumably we should still be allowed to use here

**Review by agrski — COMMENTED** [review]
I second Sherif's view that having a consumer that contains both a producer and a consumer is a weird bit of naming, although that should be able to be updated later, or we can refactor things as appropriate.

The actual change seems reasonable, although one significant question sticks out to me about multi-closing.

File: scheduler/pkg/kafka/gateway/consumer.go
Comment (cliveseldon) [reply scheduler/pkg/kafka/gateway/consumer.go:293]: will rename to producerActive

File: scheduler/pkg/kafka/gateway/consumer.go
Comment (cliveseldon) [reply scheduler/pkg/kafka/gateway/consumer.go]: will add log message

File: scheduler/pkg/kafka/gateway/consumer.go
Comment (cliveseldon) [reply scheduler/pkg/kafka/gateway/consumer.go]: will refactor

**Review by sakoush — APPROVED** [review]
LGTM

**Comment by seldondev** [comment]
[APPROVALNOTIFIER] This PR is **NOT APPROVED**

This pull-request has been approved by: *[sakoush](https://github.com/SeldonIO/seldon-core/pull/4515#pullrequestreview-1223158945 "Approved")*
To complete the [pull request process](https://git.k8s.io/community/contributors/guide/owners.md#the-code-review-process), please assign 
You can assign the PR to them by writing `/assign ` in a comment when ready.

The full list of commands accepted by this bot can be found [here](https://jenkins-x.io/v3/develop/reference/chatops/?repo=SeldonIO%2Fseldon-core).

<details open>
Needs approval from an approver in each of these files:

- **[OWNERS](https://github.com/SeldonIO/seldon-core/blob/v2/OWNERS)**

Approvers can indicate their approval by writing `/approve` in a comment
Approvers can cancel approval by writing `/approve cancel` in a comment
</details>

File: scheduler/pkg/kafka/gateway/consumer.go
Comment (agrski) [reply scheduler/pkg/kafka/gateway/consumer.go:193]: In the interests of being defensive I think it's generally good to guard against this in case someone adds a bug in the future - this is probably a fairly tricky bug to catch if it does happen.  Not a blocker for this PR mind - happy to create a little follow-up that also renames the method receivers from `kc` -> `kh`

**Pull request merged**
